### PR TITLE
ddl:create table partition value must be strictly increasing

### PR DESCRIPTION
**Pull request by ciscoxll**
## What have you changed? (mandatory)
This pr fixes two problems.
* The check partition name must be unique.
*  Check whether `less than value` is strictly increasing for each partition.
```
ErrSameNamePartition
ErrRangeNotIncreasing
ErrPartitionMaxvalue
ErrTooManyValues
ErrNotAllowedTypeInPartition
```

## What are the type of the changes (mandatory)?

Improvement

## How has this PR been tested (mandatory)?

Unit tests.

### CODE REVIEW

**Comment by ciscoxll** [comment]
@zimulala @winkyao @coocood  PTAL.

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -636,6 +637,117 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
...
+	pi := &model.PartitionInfo{
+		Type:   s.Partition.Tp,
+		Enable: ctx.GetSessionVars().EnableTablePartition,
+	}
+	if s.Partition.Expr != nil {
+		buf := new(bytes.Buffer)
+		s.Partition.Expr.Format(buf)
+		pi.Expr = buf.String()
+		if s.Partition.Tp == model.PartitionTypeRange {
+			if _, ok := s.Partition.Expr.(*ast.ColumnNameExpr); ok {
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: Only `ColumnNameExpr` is allowed?

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -636,6 +637,117 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
...
+		if err1 != nil {
+			return errors.Trace(err1)
+		}
+		piDef := model.PartitionDefinition{
+			Name:    def.Name,
+			ID:      pid,
+			Comment: def.Comment,
+		}
+		for _, expr := range def.LessThan {
+			buf := new(bytes.Buffer)
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: move it out the for-loop and reuse.
Note that `buf.String()` will make a new copy, so you don't need to worry about the memory reuse.

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -636,6 +637,117 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
...
+			expr.Format(buf)
+			piDef.LessThan = append(piDef.LessThan, buf.String())
+		}
+		pi.Definitions = append(pi.Definitions, piDef)
+	}
+	tbInfo.Partition = pi
+
+	err := checkCreatePartitionNotExists(tbInfo, tbInfo.Partition)
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: It's better to name this function `checkCreatePartitionNameUnique`

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -636,6 +637,117 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
...
+			Comment: def.Comment,
+		}
+		for _, expr := range def.LessThan {
+			buf := new(bytes.Buffer)
+			expr.Format(buf)
+			piDef.LessThan = append(piDef.LessThan, buf.String())
+		}
+		pi.Definitions = append(pi.Definitions, piDef)
+	}
+	tbInfo.Partition = pi
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: Don't set `tbInfo.Partition` until you make sure `pi` is valid.
I'm afraid you pass `tbInfo` as a reference and this change the raw data.
When error happens, it's a disaster

File: ddl/ddl_api.go [hunk 5]
```diff
@@ -636,6 +637,117 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
 	return nil
 }
 
+func checkCreatePartitionNotExists(meta *model.TableInfo, part *model.PartitionInfo) error {
+	if meta.Partition.Type == model.PartitionTypeRange {
+		newPars := part.Definitions
+		set := make(map[string]bool)
```
[the remark this answers is not in the document]
Comment (ciscoxll) [reply ddl/ddl_api.go]: Is `bool` in `map` really worth optimizing to `struct{}{}`?

[the remark this answers is not in the document]
Comment (tiancaiamao) [reply ddl/ddl_api.go]: 1. TiDB is a famous project, the code quality really matters.
2. We should have a high standard for ourselves.

File: ddl/ddl_api.go [hunk 6]
```diff
@@ -636,6 +637,113 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
 	return nil
 }
 
+func checkCreatePartitionNameUnique(part *model.PartitionInfo) error {
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: Partition name is case insensitive? Add restriction this to our wiki.

File: ddl/ddl_api.go [hunk 7]
```diff
@@ -636,6 +637,113 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
...
+			set[strings.ToLower(newPar.Name)] = struct{}{}
+		}
+	}
+	return nil
+}
+
+// checkCreatePartitionValue checks 'less than value' must be strictly increasing for each partition.
+func checkCreatePartitionValue(part *model.PartitionInfo) error {
+	if part.Type == model.PartitionTypeRange {
+		newDefs := part.Definitions
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: Well, I mean, how about this?  If `maxvalue` exists, it must be the last one:

```
defs := part.Definitions
if strings.EqualFold(defs[len(defs) - 1], "MAXVALUE") {
	defs = defs[:len(defs) - 1]
}

for i:=1; i<len(defs); i++ {
	if strings.EqualFold(defs[i].LessThan[0], "MAXVALUE") {
		errors.Trace(ErrPartitionMaxvalue)
	}
	curr := strconv.Atoi(defs[i].LessThan[0])
	prev := strconv.Atoi(defs[i-1].LessThan[0])
	if curr > prev {
		return errors.Trace(ErrRangeNotIncreasing)
	}
}
```

Comment (ciscoxll) [reply ddl/ddl_api.go]: If `maxvalue` is in the middle, `prev := strconv.Atoi(defs[i-1].LessThan[0])` will fail to convert.

Comment (ciscoxll) [reply ddl/ddl_api.go]: For example:
```
create table employees (
	id int not null,
	hired date not null
	)
	partition by range( year(hired) ) (
		partition p1 values less than (1998),
		partition p2 values less than maxvalue,
		partition p3 values less than (2001)
	);
```

File: ddl/ddl_api.go (hunk 6 above)
Comment (ciscoxll) [reply ddl/ddl_api.go]: Done.

File: ddl/ddl_api.go (hunk 7 above)
Comment (XuHuaiyu) [reply ddl/ddl_api.go]: @ciscoxll Can we use two loops:
1. the first loop is to find whether "MAXVALUE" is the last value,
2. if "MAXVALUE" is not found, we use the second loop to compare the successive digits.

**Comment by ciscoxll** [comment]
@tiancaiamao @zimulala @coocood PTAL.

File: ddl/ddl_api.go [hunk 8]
```diff
@@ -636,6 +637,107 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
...
 
+func checkCreatePartitionNameUnique(part *model.PartitionInfo) error {
+	if part.Type == model.PartitionTypeRange {
+		newPars := part.Definitions
+		set := make(map[string]struct{})
+		for _, newPar := range newPars {
+			if _, ok := set[newPar.Name]; ok {
+				return ErrSameNamePartition.GenByArgs(newPar.Name)
+			}
+			set[newPar.Name] = struct{}{}
```
Comment (shenli) [inline ddl/ddl_api.go]: partition name is case-insensitive.

File: ddl/ddl_api.go [hunk 9]
```diff
@@ -636,6 +637,108 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
...
+			if nextRangeValue <= currentRangeValue {
+				return errors.Trace(ErrRangeNotIncreasing)
+			}
+			currentRangeValue = nextRangeValue
+		}
+	}
+	return nil
+}
+
+func buildCreateTablePartitionInfo(ctx sessionctx.Context, d *ddl, s *ast.CreateTableStmt, cols []*table.Column) (*model.PartitionInfo, error) {
```
Comment (zimulala) [inline ddl/ddl_api.go]: Could we move these function to the file of "ddl/partition.go"? There is a lot of code in the current file.

File: ddl/ddl_api.go [hunk 10]
```diff
@@ -636,6 +637,108 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
...
+	}
+	if s.Partition.Expr != nil {
+		buf := new(bytes.Buffer)
+		s.Partition.Expr.Format(buf)
+		pi.Expr = buf.String()
+		if s.Partition.Tp == model.PartitionTypeRange {
+			for _, col := range cols {
+				name := strings.Replace(col.Name.String(), ".", "`.`", -1)
+				if _, ok := s.Partition.Expr.(*ast.ColumnNameExpr); ok {
+					// TODO check that the expression returns an integer.
```
Comment (zimulala) [inline ddl/ddl_api.go]: s/TODO/TODO:

File: ddl/ddl_api.go [hunk 11]
```diff
@@ -636,6 +637,108 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
...
+				}
+				if _, ok := s.Partition.Expr.(ast.ExprNode); ok {
+					if !(col.Tp == mysql.TypeLong || col.Tp == mysql.TypeLonglong) && fmt.Sprintf("`%s`", name) == pi.Expr {
+						return nil, errors.Trace(ErrNotAllowedTypeInPartition.GenByArgs(pi.Expr))
+					}
+				}
+			}
+
```
Comment (zimulala) [inline ddl/ddl_api.go]: Remove this useless line.

File: ddl/ddl_api.go [hunk 12]
```diff
@@ -636,6 +637,108 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
...
+		s.Partition.Expr.Format(buf)
+		pi.Expr = buf.String()
+		if s.Partition.Tp == model.PartitionTypeRange {
+			for _, col := range cols {
+				name := strings.Replace(col.Name.String(), ".", "`.`", -1)
+				if _, ok := s.Partition.Expr.(*ast.ColumnNameExpr); ok {
+					// TODO check that the expression returns an integer.
+				}
+				if _, ok := s.Partition.Expr.(ast.ExprNode); ok {
+					if !(col.Tp == mysql.TypeLong || col.Tp == mysql.TypeLonglong) && fmt.Sprintf("`%s`", name) == pi.Expr {
```
Comment (zimulala) [inline ddl/ddl_api.go]: s/`!(col.Tp == mysql.TypeLong || col.Tp == mysql.TypeLonglong)`/`col.Tp != mysql.TypeLong && col.Tp != mysql.TypeLonglong`?

File: ddl/ddl_api.go [hunk 13]
```diff
@@ -636,6 +637,108 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
...
+		}
+	} else if s.Partition.ColumnNames != nil {
+		pi.Columns = make([]model.CIStr, 0, len(s.Partition.ColumnNames))
+		for _, cn := range s.Partition.ColumnNames {
+			pi.Columns = append(pi.Columns, cn.Name)
+		}
+	}
+	for _, def := range s.Partition.Definitions {
+		// TODO: generate multiple global ID for paritions.
+		pid, err1 := d.genGlobalID()
```
Comment (zimulala) [inline ddl/ddl_api.go]: s/err1/err

File: ddl/ddl_api.go (hunk 12 above)
Comment (ciscoxll) [reply ddl/ddl_api.go]: @zimulala `col.Tp != mysql.TypeLong && col.Tp != mysql.TypeLonglong` the logic is wrong.

Comment (ciscoxll) [reply ddl/ddl_api.go]: @zimulala It's not easy to understand if you write it this way.
```
col.Tp != mysql.TypeLong || col.Tp != mysql.TypeLonglong && fmt.Sprintf("`%s`", name) == pi.Expr
```

**Comment by ciscoxll** [comment]
@tiancaiamao @zimulala PTAL.

File: ddl/partition.go [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+			piDef.LessThan = append(piDef.LessThan, buf.String())
+			buf.Reset()
+		}
+		pi.Definitions = append(pi.Definitions, piDef)
+	}
+	return pi, nil
+}
+
+func checkCreatePartitionNameUnique(part *model.PartitionInfo) error {
+	if part.Type != model.PartitionTypeRange {
```
Comment (zimulala) [inline ddl/partition.go]: you can put this check to line 847 in `ddl_api.go`. Then you can remove this check here and in function of `checkCreatePartitionValue`.

File: ddl/partition.go [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+		}
+		if strings.EqualFold(defs[len(defs)-1].LessThan[0], "MAXVALUE") {
+			defs = defs[:len(defs)-1]
+		}
+
+		for i := 1; i < len(defs); i++ {
+			if strings.EqualFold(defs[i].LessThan[0], "MAXVALUE") && i != len(defs)-1 {
+				return errors.Trace(ErrPartitionMaxvalue)
+			}
+			nextRangeValue, err := strconv.Atoi(defs[i].LessThan[0])
```
Comment (zimulala) [inline ddl/partition.go]: I think `nextRangeValue` is currentRangeValue, and `currentRangeValue` is prevRangeValue.

File: ddl/partition.go [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		piDef := model.PartitionDefinition{
+			Name:    def.Name,
+			ID:      pid,
+			Comment: def.Comment,
+		}
+		buf := new(bytes.Buffer)
+		for _, expr := range def.LessThan {
```
[the remark this answers is not in the document]
Comment (ciscoxll) [reply ddl/partition.go]: We support `range columns` statement, for example @zimulala .
```
CREATE TABLE t (a int, b int, c char, d int) 
PARTITION BY RANGE COLUMNS(a,d,c) (
 	PARTITION p0 VALUES LESS THAN (5,10,'ggg'),
 	PARTITION p1 VALUES LESS THAN (10,20,'mmm'),
 	PARTITION p2 VALUES LESS THAN (15,30,'sss'),
        PARTITION p3 VALUES LESS THAN (50,MAXVALUE,MAXVALUE)
    );
```
So can't remove this `for`.

[the remark this answers is not in the document]
Comment (zimulala) [reply ddl/partition.go]: Got it.

**Comment by ciscoxll** [comment]
tiancaiamao @zimulala PTAL

File: ddl/partition.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (zimulala) [reply ddl/partition.go]: Could we add a test for it？

[the remark this answers is not in the document]
Comment (ciscoxll) [reply ddl/partition.go]: @zimulala PTAL

**Comment by ciscoxll** [comment]
@tiancaiamao PTAL

**Comment by ciscoxll** [comment]
@shenli  @tiancaiamao PTAL

File: ddl/ddl_api.go [hunk 17]
```diff
@@ -838,50 +838,22 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
...
-					}
-				}
-			}
-		} else if s.Partition.ColumnNames != nil {
-			pi.Columns = make([]model.CIStr, 0, len(s.Partition.ColumnNames))
-			for _, cn := range s.Partition.ColumnNames {
-				pi.Columns = append(pi.Columns, cn.Name)
-			}
+
+	part, err := buildCreateTablePartitionInfo(ctx, d, s, cols)
```
Comment (shenli) [inline ddl/ddl_api.go]: s/buildCreateTablePartitionInfo/buildTablePartitionInfo/

File: ddl/ddl_api.go [hunk 18]
```diff
@@ -838,50 +838,22 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
...
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if part != nil && part.Type == model.PartitionTypeRange {
+		tbInfo.Partition = part
+		err = checkCreatePartitionNameUnique(tbInfo.Partition)
```
Comment (shenli) [inline ddl/ddl_api.go]: Can we merge the two checks into one function?

File: ddl/partition.go [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/table"
+)
+
+func buildCreateTablePartitionInfo(ctx sessionctx.Context, d *ddl, s *ast.CreateTableStmt, cols []*table.Column) (*model.PartitionInfo, error) {
```
Comment (shenli) [inline ddl/partition.go]: Add comments in the functions.

File: ddl/partition.go [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+	if s.Partition == nil {
+		return nil, nil
+	}
+	pi := &model.PartitionInfo{
+		Type:   s.Partition.Tp,
+		Enable: ctx.GetSessionVars().EnableTablePartition,
+	}
+	if s.Partition.Expr != nil {
+		buf := new(bytes.Buffer)
+		s.Partition.Expr.Format(buf)
```
Comment (shenli) [inline ddl/partition.go]: Why need to format the expr?

File: ddl/partition.go [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+				buf.Reset()
+			}
+			pi.Definitions = append(pi.Definitions, piDef)
+		}
+	}
+	return pi, nil
+}
+
+func checkCreatePartitionNameUnique(part *model.PartitionInfo) error {
+	newPars := part.Definitions
```
Comment (shenli) [inline ddl/partition.go]: Do we need the `newPars`?

File: ddl/partition.go [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+	set := make(map[string]struct{})
```
Comment (shenli) [inline ddl/partition.go]: s/set/partNames/

File: ddl/partition.go [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+		set[strings.ToLower(newPar.Name)] = struct{}{}
+	}
+	return nil
+}
+
+// checkCreatePartitionValue checks whether `less than value` is strictly increasing for each partition.
+func checkCreatePartitionValue(part *model.PartitionInfo) error {
+	defs := part.Definitions
+	rangeValue := defs[0].LessThan[0]
+	if strings.EqualFold(rangeValue, "MAXVALUE") && len(defs) > 1 {
```
Comment (shenli) [inline ddl/partition.go]: Create a constant for MAXVALUE.

Comment (shenli) [reply ddl/partition.go]: What is this check used for?

File: ddl/partition.go [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+	prevRangeValue, err := strconv.Atoi(rangeValue)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if strings.EqualFold(defs[len(defs)-1].LessThan[0], "MAXVALUE") {
+		defs = defs[:len(defs)-1]
+	}
+
+	for i := 1; i < len(defs); i++ {
+		if strings.EqualFold(defs[i].LessThan[0], "MAXVALUE") && i != len(defs)-1 {
```
Comment (shenli) [inline ddl/partition.go]: Can not understand this. You have already removed the last  MAXVALUE (if there is) at line 113. Can we just return error without checking `i != len(defs)-1`?

File: ddl/table.go [hunk 25]
```diff
@@ -437,15 +450,15 @@ func updatePartitionInfo(partitionInfo *model.PartitionInfo, tblInfo *model.Tabl
 
 func checkPartitionNotExists(meta *model.TableInfo, part *model.PartitionInfo) error {
 	oldPars, newPars := meta.Partition.Definitions, part.Definitions
-	set := make(map[string]bool)
+	set := make(map[string]struct{})
```
Comment (shenli) [inline ddl/table.go]: s/set/partNames/

File: ddl/table.go [hunk 26]
```diff
@@ -437,15 +450,15 @@ func updatePartitionInfo(partitionInfo *model.PartitionInfo, tblInfo *model.Tabl
...
 	for _, oldPar := range oldPars {
-		set[strings.ToLower(oldPar.Name)] = true
+		set[strings.ToLower(oldPar.Name)] = struct{}{}
```
Comment (shenli) [inline ddl/table.go]: partition names are case-insensitive.

File: ddl/ddl_api.go [hunk 27]
```diff
@@ -838,50 +838,22 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
...
-				pi.Columns = append(pi.Columns, cn.Name)
-			}
+
+	part, err := buildCreateTablePartitionInfo(ctx, d, s, cols)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if part != nil && part.Type == model.PartitionTypeRange {
+		tbInfo.Partition = part
```
Comment (zimulala) [inline ddl/ddl_api.go]: Do we need to do this for other partition type？

File: ddl/table.go (hunk 26 above)
Comment (shenli) [reply ddl/table.go]: Ignore it please.

**Comment by ciscoxll** [comment]
@tiancaiamao @zimulala @shenli  PTAL.

**Comment by ciscoxll** [comment]
@zimulala @shenli  PTAL.

File: ddl/partition.go (hunk 20 above)
Comment (ciscoxll) [reply ddl/partition.go]: We can get the `expr` type directly without asserting.

**Comment by ciscoxll** [comment]
@tiancaiamao PTAL.

File: ddl/partition.go [hunk 28]
```diff
@@ -0,0 +1,142 @@
...
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/table"
+)
+
+var (
+	maxValue = "MAXVALUE"
```
Comment (zimulala) [inline ddl/partition.go]: s/maxValue/partitionMaxValue ?

File: ddl/partition.go [hunk 29]
```diff
@@ -0,0 +1,142 @@
...
+)
+
+// buildTablePartitionInfo build partition info and checks for some errors.
```
Comment (zimulala) [inline ddl/partition.go]: s/build/builds

File: ddl/ddl_api.go [hunk 30]
```diff
@@ -838,50 +838,22 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
...
-				expr.Format(buf)
-				piDef.LessThan = append(piDef.LessThan, buf.String())
-			}
-			pi.Definitions = append(pi.Definitions, piDef)
+		err = checkCreatePartitionValue(part)
+		if err != nil {
+			return errors.Trace(err)
 		}
-		tbInfo.Partition = pi
+		tbInfo.Partition = part
```
Comment (zimulala) [inline ddl/ddl_api.go]: If we only get the partition of type "PartitionTypeRange"，why do we check it so many places？

**Comment by ciscoxll** [comment]
@zimulala @tiancaiamao  PTAL.

File: ddl/partition.go [hunk 31]
```diff
@@ -0,0 +1,142 @@
...
+			}
+		}
+	} else if s.Partition.ColumnNames != nil {
+		pi.Columns = make([]model.CIStr, 0, len(s.Partition.ColumnNames))
+		for _, cn := range s.Partition.ColumnNames {
+			pi.Columns = append(pi.Columns, cn.Name)
+		}
+	}
+	for _, def := range s.Partition.Definitions {
+		// TODO: generate multiple global ID for paritions.
```
Comment (shenli) [inline ddl/partition.go]: Why ned to `generate multiple global ID`?

File: model/model.go [hunk 32]
```diff
@@ -287,6 +287,7 @@ type PartitionDefinition struct {
 	ID       int64    `json:"id"`
 	Name     string   `json:"name"`
 	LessThan []string `json:"less_than"`
+	MaxValue bool     `json:"max_value"`
```
Comment (shenli) [inline model/model.go]: s/MaxValue/IsMaxValue/

File: ddl/partition.go [hunk 33]
```diff
@@ -0,0 +1,142 @@
...
+func checkCreatePartitionValue(part *model.PartitionInfo) error {
+	defs := part.Definitions
+	if len(defs) <= 1 {
+		return nil
+	}
+	if defs[len(defs)-1].MaxValue {
+		defs = defs[:len(defs)-1]
+	}
+	for i := 1; i < len(defs); i++ {
+		if defs[i-1].MaxValue || defs[i].MaxValue {
```
Comment (shenli) [inline ddl/partition.go]: Why need to check defs[i-1].MaxValue?

File: ddl/partition.go [hunk 34]
```diff
@@ -0,0 +1,142 @@
...
+			return errors.Trace(err)
+		}
+		currentRangeValue, err := strconv.Atoi(defs[i].LessThan[0])
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if currentRangeValue <= prevRangeValue {
+			return errors.Trace(ErrRangeNotIncreasing)
+		}
+		prevRangeValue = currentRangeValue
```
Comment (shenli) [inline ddl/partition.go]: if you have line 128, you do not need this line. I suggest treating the first item as a special case, so you do not need to check defs[i-1].MaxValue and evaluate prevRangeValue in each loop.

File: ddl/partition.go [hunk 35]
```diff
@@ -0,0 +1,142 @@
...
+		s.Partition.Expr.Format(buf)
+		pi.Expr = buf.String()
+		if s.Partition.Tp == model.PartitionTypeRange {
+			for _, col := range cols {
+				name := strings.Replace(col.Name.String(), ".", "`.`", -1)
+				if _, ok := s.Partition.Expr.(*ast.ColumnNameExpr); ok {
+					// TODO: check that the expression returns an integer.
+				}
+				if _, ok := s.Partition.Expr.(ast.ExprNode); ok {
+					if col.Tp != mysql.TypeLong && col.Tp != mysql.TypeLonglong && fmt.Sprintf("`%s`", name) == pi.Expr {
```
Comment (shenli) [inline ddl/partition.go]: Add comment for this if condition.

File: ddl/partition.go [hunk 36]
```diff
@@ -0,0 +1,142 @@
...
+			ID:      pid,
+			Comment: def.Comment,
+		}
+
+		if s.Partition.Tp == model.PartitionTypeRange {
+			if s.Partition.ColumnNames == nil && len(def.LessThan) != 1 {
+				return nil, ErrTooManyValues.GenByArgs("RANGE")
+			}
+			buf := new(bytes.Buffer)
+			for _, expr := range def.LessThan {
```
Comment (shenli) [inline ddl/partition.go]: Add comment for this.

File: ddl/partition.go [hunk 37]
```diff
@@ -0,0 +1,142 @@
...
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/table"
+)
+
+var (
```
Comment (tiancaiamao) [inline ddl/partition.go]: const

File: model/model.go (hunk 32 above)
Comment (tiancaiamao) [reply model/model.go]: What a terrible idea to have this field in `PartitionDefinition`.
It's redundancy information, it's noise.

**Comment by ciscoxll** [comment]
@shenli  @tiancaiamao PTAL.

File: ddl/partition.go (hunk 31 above)
Comment (ciscoxll) [reply ddl/partition.go]: Please @coocood answer the question.

File: ddl/partition.go [hunk 38]
```diff
@@ -0,0 +1,149 @@
...
+func checkCreatePartitionValue(part *model.PartitionInfo) error {
+	defs := part.Definitions
+	if len(defs) <= 1 {
+		return nil
+	}
+
+	if strings.EqualFold(defs[len(defs)-1].LessThan[0], partitionMaxValue) {
+		defs = defs[:len(defs)-1]
+	}
+	if strings.EqualFold(defs[0].LessThan[0], partitionMaxValue) {
```
Comment (tiancaiamao) [inline ddl/partition.go]: `if len(defs) > 0 && ...`

Comment (ciscoxll) [reply ddl/partition.go]: @tiancaiamao Line 117 is already included.
```
if len(defs) <= 1 {
	return nil
}
```

**Comment by ciscoxll** [comment]
@shenli @tiancaiamao PTAL.

**Comment by tiancaiamao** [comment]
LGTM

**Comment by ciscoxll** [comment]
@shenli @zimulala  PTAL.

**Comment by ciscoxll** [comment]
@shenli @zimulala PTAL.

File: ddl/ddl_db_test.go [hunk 39]
```diff
@@ -1578,6 +1580,83 @@ func (s *testDBSuite) TestCreateTableWithPartition(c *C) {
...
+	s.testErrorCode(c, sql4, tmysql.ErrPartitionMaxvalue)
+
+	sql5 := `create table employees (
+	id int not null,
+	hired int not null
+	)
+	partition by range( hired ) (
+		 PARTITION p0 VALUES LESS THAN (6 , 10)
+	);`
+	s.testErrorCode(c, sql5, tmysql.ErrTooManyValues)
```
Comment (zimulala) [inline ddl/ddl_db_test.go]: Could we add the test as follows：
```
CREATE TABLE rc (
    a INT NOT NULL,
    b INT NOT NULL,
	c INT NOT NULL
)
PARTITION BY RANGE COLUMNS(a,b,c) (
    PARTITION p0 VALUES LESS THAN (10,5,1),
    PARTITION p2 VALUES LESS THAN (50,MAXVALUE,10),
    PARTITION p3 VALUES LESS THAN (65,30,13),
    PARTITION p4 VALUES LESS THAN (MAXVALUE,30,40)
);
```

**Comment by ciscoxll** [comment]
@zimulala PTAL.

**Comment by ciscoxll** [comment]
@shenli @zimulala PTAL.

**Comment by ciscoxll** [comment]
@shenli PTAL.

File: ddl/partition.go [hunk 40]
```diff
@@ -0,0 +1,158 @@
...
+
+		currentRangeValue, err := strconv.Atoi(defs[i].LessThan[0])
+		if err != nil {
+			return ErrNotAllowedTypeInPartition.GenByArgs(defs[i].LessThan[0])
+		}
+
+		if i == 0 {
+			continue
+		}
+		prevRangeValue, err := strconv.Atoi(defs[i-1].LessThan[0])
```
Comment (shenli) [inline ddl/partition.go]: Can you save the prevRangeValue at the end of each loop?

**Comment by ciscoxll** [comment]
@shenli PTAL.

**Comment by ciscoxll** [comment]
@shenli PTAL.

**Pull request merged**